### PR TITLE
kv: disallow parallel commit of batch with read-only requests

### LINKED ISSUE #44315: kv: unexpected REASON_TXN_COMMITTED from CommitInBatch
The PR to add support to kvnemeses for exercising txn.CommitInBatch (#44229) will occasionally flake on a `TransactionStatusError: already committed (REASON_TXN_COMMITTED)` error. At Nathan's request, I checked the txn.Epoch after one of these failures and it was `0` indicating no restarts. Perhaps this is missing a retry somewhere?

```
  db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
    txn.Put(ctx, /Table/50/"e14094f1", v-15) // nil
    b := &Batch{}
    b.Get(ctx, /Table/50/"0b10b00b") // (nil, TransactionStatusError: already committed (REASON_TXN_COMMITTED))
    b.Get(ctx, /Table/50/"0b10b00b") // (nil, TransactionStatusError: already committed (REASON_TXN_COMMITTED))
    txn.CommitInBatch(ctx, b) // TransactionStatusError: already committed (REASON_TXN_COMMITTED)
    return nil
  }) // TransactionStatusError: already committed (REASON_TXN_COMMITTED)
```

Repro: Check out #44229 and run the following. It tends to fail within a minute on my laptop.
```
make stress PKG=./pkg/kv/kvnemeses TESTS=TestKVNemesesSingleNode TESTTIMEOUT=45s TESTFLAGS='-v -show-logs' STRESSFLAGS='-stderr -maxfails 1'
```

Found with kvnemeses

### PR DESCRIPTION
**Pull request by nvb**
Fixes #44315.

This commit fixes an issue where KV batches with read-only requests were allowed
to be issued as parallel commits. This should not have been allowed because the
outcome of these read-only requests, notably Get and Scan requests, is not taken
into consideration by the status resolution process for STAGING transactions.

This would have caused serious atomicity issues if parallel commit batches
(non-1PC committing batches) were issued with read-only requests by users of the
KV API. Luckily, in practice that wouldn't actually happen as we never issue
batches like that so there's no concern about production clusters hitting this
bug. Still, this is a very good catch as this was a serious footgun and fixing
this hardens the KV API. We should also still backport this to v19.2 just to be
safe.

Release note: None

### CODE REVIEW

**Comment by danhhz** [comment]
I could rubber stamp this, the change looks fine and like it's doing what it says it does, but would we prefer a real review?

**Review by andreimatei — APPROVED** [review]
LGTM, but to be honest I'm not sure what semantics we want for reads in the commit batch. I think you can argue for committing even if these reads fail. But we can punt until we have a better error recovery story.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44608#-:-M-BT5vg7EDOWQAvuTcl:b-bb3vi8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @danhhz, and @nvanbenschoten)

___
*[pkg/kv/txn_interceptor_committer.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44608#-M-BSKGjEB0Eb_4RC6Uf:-M-BSKGjEB0Eb_4RC6Ug:bjft380) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66c6c3709133041bca05c37c62766b22a8e6371/pkg/kv/txn_interceptor_committer.go#L279)):*
> ```Go
> // canBatchCommitInParallel determines whether the batch can issue its
> // committing EndTxn in parallel with the rest of its requests.
> func (tc *txnCommitter) canBatchCommitInParallel(
> ```

The new name doesn't mean much to me. How about just `canCommitInParallel`?
And I'd rephrase the comment too; it's not just about committing in parallel with the other requests in this batch, but also with the in-flight writes.

**Review by nvb — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44608#-:-M-CiICQDkz3wgH6Ujbe:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @danhhz)

___
*[pkg/kv/txn_interceptor_committer.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44608#-M-BSKGjEB0Eb_4RC6Uf:-M-CiH_p0OpvcYuC_0MC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66c6c3709133041bca05c37c62766b22a8e6371/pkg/kv/txn_interceptor_committer.go#L279)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The new name doesn't mean much to me. How about just `canCommitInParallel`?
And I'd rephrase the comment too; it's not just about committing in parallel with the other requests in this batch, but also with the in-flight writes.
</blockquote></details>

Done.

**Comment by nvanbenschoten** [comment]
`no space left on device` then `TestRandomKeyAndTimestampExport` flake (https://github.com/cockroachdb/cockroach/issues/44589).

bors r+

**Comment by nvanbenschoten** [comment]
Hello bors?

```
{{:badmatch,
  {:error, :push, 422,
   "{\"message\":\"Required status check \\\"license/cla\\\" is expected.\",\"documentation_url\":\"https://help.github.com/articles/about-protected-branches\"}"}},
 [
   {BorsNG.Worker.Batcher, :complete_batch, 3,
    [file: 'lib/worker/batcher.ex', line: 475]},
   {BorsNG.Worker.Batcher, :maybe_complete_batch, 1,
    [file: 'lib/worker/batcher.ex', line: 456]},
   {BorsNG.Worker.Batcher, :handle_cast, 2,
    [file: 'lib/worker/batcher.ex', line: 83]},
   {:gen_server, :try_dispatch, 4,
    [file: 'gen_server.erl', line: 637]},
   {:gen_server, :handle_msg, 6,
    [file: 'gen_server.erl', line: 711]},
   {:proc_lib, :init_p_do_apply, 3,
    [file: 'proc_lib.erl', line: 249]}
 ]}
```

bors r+
